python/cornserve/services/resource_manager/manager.py

```diff
@@ -186,7 +190,34 @@ async def init() -> ResourceManager:
                     await asyncio.gather(*cleanup_coros, return_exceptions=True)
                 raise RuntimeError(f"Failed to spawn {failed} sidecar pods: {spawn_results}")
 
-            # TODO (Jeff): CheckHealth with Timeout
+            if failed:
+                await cleanup()
+            else:
+
+                async def wait_for_online(rank: int) -> None:
+                    """Wait for sidecar with `rank` to be online."""
+                    while True:
+                        try:
+                            async with grpc.aio.insecure_channel(grpc_url_from_rank(rank)) as channel:
+                                req = sidecar_pb2.CheckHealthRequest()
+                                stub = sidecar_pb2_grpc.SidecarStub(channel)
+                                res = await stub.CheckHealth(req)
+                                if res.status == sidecar_pb2.HealthStatus.HEALTH_ALL_GOOD:
+                                    logger.info("Sidecar %d is online", rank)
+                                    return
+                                await asyncio.sleep(3)
+                        except Exception:
+                            await asyncio.sleep(3)
+
+                coros = [wait_for_online(gpu.global_rank) for gpu in gpus]
+                try:
+                    async with asyncio.timeout(SidecarLaunchInfo.DEFAULT_LAUNCH_TIMEOUT):
+                        await asyncio.gather(*coros)
+                except TimeoutError:
+                    logger.error("Timed out waiting for sidecars to come online")
+                    await cleanup()
+                logger.info("All sidecars are online")
+
```

Comment:
Failure case eventually reaches L219 and successfully begins the RM!?

Reply:
Oh, `cleanup` raises an error. This is not a good pattern. It should just clean up and return. Then, on the next line, you should write `raise RuntimeError(...) from e` where `e` is `TimeoutError` or something else that caused the `RuntimeError`.